### ISSUE TITLE
Feature request: Automatically close connections that haven't been used for a while

### ISSUE THREAD

**Issue by sosthene-nitrokey** [algesten/ureq#772]
Hi,

We are encountering multiple cases in our customer's networks where idle ureq connections are kept alive by ureq indefinitely, but are killed silently by an intermediate firewall. This leads to the next request (sometimes tens of minutes after that) failing after a timeout (since the firewall doesn't even send a `FIN` packet, when closing the connection or a `RST` packet when receiving a new packet).

Our current solution is to manually implement retries, but that is not ideal because of the timeout, and if there are multiple connections in the pool it might take multiple retries to go through all of them.

The solution I suggest: When building an agent, accept a configuration `connection_stale_after` accepting a duration.
When making a request, and getting new connections, if an existing connection's last use is older than the duration `connection_stale_after`, drop the connection.

**Comment by algesten** [comment]
@sosthene-nitrokey thanks for reaching out!

We've been conservative with spawning new threads instead thinking that if people do operations using an agent often enough, the stale connections are cleaned up. Do you think we should spawn a pool freshness thread?

**Comment by sosthene-nitrokey** [comment]
I don't think this should be done by background thread (by default at least. Having this available as opt-in would be great).
I having the "stale" connections be dropped when a request is initiated would be good enough for a lot of use cases.

**Comment by algesten** [comment]
>  I having the "stale" connections be dropped when a request is initiated would be good enough for a lot of use cases.

Uhm. I thought that is what we are doing? 😄

**Comment by sosthene-nitrokey** [comment]
From what I see connections that have IO failures are dropped, but they are dropped only after the failure. I want to be able to pre-emptively close them.

This is because in our case the connection is closed by a firewall that does not send `RST` or `FIN`, so the only thing that happens is the timeout, on the read call, not even on the write call for sending the prelude.

I want to not even try to write to a connection it if I know it has been closed by the firewall due to having been stale for a long time to avoid having to deal with the timeouts.

**Comment by sosthene-nitrokey** [comment]
Oh, I think I see what's happening.

`send_prelude` does not flush, so the `write` is buffered by the kernel and `send_prelude` succeed, even though the data is only ever sent into a silent firewall.

Then the call to `read` times out (or the write of the body I'm not sure), so the error bubbles up to us, instead of leading to the immediate  retry mechanism in `unit::connect_inner` which only errors for the prelude send call.

I'm not even sure that flushing there would be desirable because it would introduce a delay before sending the beginning of the body.

**Comment by sosthene-nitrokey** [comment]
Oh wait there is no way to actually flush a TcpStream...

**Comment by sosthene-nitrokey** [comment]
Otherwise adding support for TCP keepalive would also help us fix our issue, since the keepalive would likely detect the connection being dead early, without the need for a background thread.

**Comment by algesten** [comment]
Yeah. That's one of those irritating things in Rust, right?

**Comment by sosthene-nitrokey** [comment]
We have implemented the features we needed in a fork of ureq, and are planning to use them in a future release of our `neths-pkcs11` crate.

We're on 2.9.1 so it's built on top of that, but I will check out if we can use the 3.0 branch so that we can suggest them as PRs:

- https://github.com/Nitrokey/ureq/pull/2
- https://github.com/Nitrokey/ureq/pull/1

**Comment by sosthene-nitrokey** [comment]
Oh, this feature is actually already available in 3.0 as `max_idle_age`

**Comment by algesten** [comment]
Yeah, but it builds on that the pool is used, or they are not cleaned up. We could make some optional thread forking.

**Comment by algesten** [comment]
@sosthene-nitrokey can we close this?

**Comment by sosthene-nitrokey** [comment]
Yes this is already built into the 3.0 release.
